### PR TITLE
subtile intersection via the tiles object

### PR DESCRIPTION
**Pull request by kevinkreiser**
basically we need a way to cut up the tiles into even smaller pieces to efficiently tell where certain roads are, please read: https://github.com/valhalla/loki/issues/72

i'm not going to merge this until i have all my tests working. once that is complete ill have to turn back on the binning stuff in `mjolnir`.

after that ill need to come back and flesh out the radius intersection and tests.

also note that i deleted clipper2 and moved its methods to aabb2. if we like i can pr that seprately as it kind of confuses this diff. let me know!

### CODE REVIEW

**Comment by dgearhart** [comment]
are we waiting on test for this?

**Comment by kevinkreiser** [comment]
well its not used yet so we could merge it without issue. but im not going to hook it up to graphvalidator before i pr with fully working tests (they are commented out currently)

**Comment by kevinkreiser** [comment]
sorry for the long stint... this is ready for review

File: src/midgard/aabb2.cc [hunk 1]
```diff
@@ -199,6 +212,159 @@ bool AABB2<coord_t>::Intersects(const coord_t& c, float r) const {
...
+  float u_dist = std::numeric_limits<float>::infinity(), v_dist = std::numeric_limits<float>::infinity(), d;
+  for(const auto& intersection : intersections) {
+    if(need_u && (d = u.DistanceSquared(intersection)) < u_dist) { u = intersection; u_dist = d; }
+    if(need_v && (d = v.DistanceSquared(intersection)) < v_dist) { v = intersection; v_dist = d; }
+  }
+  //are we inside now?
+  return intersections.size();
+}
+
+// Clips the input set of vertices to the specified boundary.  Uses a
```
Comment (kevinkreiser) [inline src/midgard/aabb2.cc]: from here down was moved from clipper2

File: src/midgard/tiles.cc [hunk 2]
```diff
@@ -1,5 +1,75 @@
...
+      else { x0 += sx; y0 += sy; }
+      //mark this pixel
+      auto o = set_pixel(x0, y0);
+      if(outside == false && o == true)
+        return;
+      outside = o;
+    }
+  }
+
+  void bresenham_circle(int32_t x0, int32_t y0, int32_t r, const std::function<bool (int32_t, int32_t)>& set_pixel) {
```
Comment (kevinkreiser) [inline src/midgard/tiles.cc]: not  yet tested

File: src/midgard/util.cc [hunk 3]
```diff
@@ -310,5 +310,52 @@ container_t resample_spherical_polyline(const container_t& polyline, double reso
...
 
+//Return the intersection of two infinite lines if any
+template <class coord_t>
+bool intersect(const coord_t& u, const coord_t& v, const coord_t& a, const coord_t& b, coord_t& i) {
+  auto uv_xd = u.first - v.first;
+  auto uv_yd = u.second - v.second;
+  auto ab_xd = a.first - b.first;
+  auto ab_yd = a.second - b.second;
+  auto d_cross = uv_xd*ab_yd - ab_xd*uv_yd;
+  if(std::abs(d_cross) < 1e-5)
```
Comment (dnesbitt61) [inline src/midgard/util.cc]: Maybe add some comments - assume this is to check if parallel?

**Comment by dnesbitt61** [comment]
:shipit:

File: src/midgard/util.cc (hunk 3 above)
Comment (kevinkreiser) [inline src/midgard/util.cc]: yes and will do!

**Pull request merged**